File: Data Structures & Algorithms/islands-and-treasure/submission-1.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 0:
                    stack = [(i, j, 0)]
                    seen = [[False for j in range(len(grid[0]))] for i in range(len(grid))]

                    while stack:
                        cell = stack.pop()
                        seen[cell[0]][cell[1]] = True
                        grid[cell[0]][cell[1]] = min(grid[cell[0]][cell[1]], cell[2])

                        for k in range(4):
                            x = [1, 0, -1, 0]
                            y = [0, -1, 0, 1]
                            row = cell[0] + y[k]
                            col = cell[1] + x[k]

                            if not ((0 <= row < len(grid)) and (0 <= col < len(grid[0]))):
                                continue

                            newDistance = cell[2] + 1

                            if (grid[row][col] > 0) and (newDistance < grid[row][col]):
                                stack.append((row, col, newDistance))
                    

#[inf,inf,inf]
#[inf, -1,inf]
#[0  ,inf,inf]

#[2, 3, 4]
#[1,-1, 5]
#[0, 1, 6]]

#sol = Solution()
#sol.islandsAndTreasure([[1, 0, 4], [2, 3, 4]])

File: Data Structures & Algorithms/search-for-word/submission-1.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        struct Tile {
            int row = 0;
            int col = 0;
            //int srcDir = 0; // (-1 = N/A, 0 = left, 1 = up, 2 = right, 3 = down)
            set<vector<int>> seen;
            int idx = 0;
        };

        int height = board.size();
        int width = board[0].size();
        int wordLen = word.size();
        // row, col, src dir (-1 = N/A, 0 = left, 1 = up, 2 = right, 3 = down), idx
        stack<Tile> st;
        vector<int> dR = {0, -1, 0, 1};
        vector<int> dC = {-1, 0, 1, 0};
        vector<int> oppositeDirs = {2, 3, 0, 1};
 
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                if (board[i][j] == word[0]) st.push({i, j, {}, 0});    

                while (!st.empty()) {
                    Tile tile = st.top();
                    st.pop();
                    if (tile.idx == wordLen - 1) return true;
                    int newIdx = tile.idx + 1;
                    tile.seen.insert({tile.row, tile.col});

                    for (int k = 0; k < 4; k++) {
                        int newR = tile.row + dR[k];
                        int newC = tile.col + dC[k];
                        vector<int> newCoords {newR, newC};
                        bool rValid = newR >= 0 && newR < height;
                        bool cValid = newC >= 0 && newC < width;
                        bool valid = rValid && cValid;

                        if (valid && board[newR][newC] == word[newIdx] && !tile.seen.contains(newCoords)) {
                            st.push({newR, newC, tile.seen, newIdx});
                        }
                    }
                }
            }
        }

        return false;
    }
};